### PR TITLE
DrawingView fixes

### LINKED ISSUE #108: [Proposal] DrawingView
# DrawingView

* [x] Proposed
* [x] Prototype
* [x] Implementation
  * [x] iOS Support
  * [x] Android Support
  * [x] macOS Support
  * [x] Windows Support
* [x] Unit Tests
* [x] Sample
* [x] Documentation

## Summary
[summary]: #summary

Allows you to draw one or multiple lines on a canvas

## Detailed Design
[design]: #detailed-design

### DrawingView.shared.cs

```cs
public class DrawingView : View
{
  public static readonly BindableProperty ClearOnFinishProperty;
  public static readonly BindableProperty MultiLineModeProperty;
  public static readonly BindableProperty LinesProperty;
  public static readonly BindableProperty DrawingLineCompletedCommandProperty;
  public static readonly BindableProperty DefaultLineColorProperty;
  public static readonly BindableProperty DefaultLineWidthProperty;
  
  public event EventHandler<DrawingLineCompletedEventArgs>? DrawingLineCompleted;
  
  public Color DefaultLineColor { get; set; }
  public float DefaultLineWidth { get; set; }
  public ICommand? DrawingLineCompletedCommand { get; set; }
  public ObservableCollection<Line> Lines { get; set; }
  public bool MultiLineMode { get; set; }
  public bool ClearOnFinish { get; set; }
  
  public Stream GetImageStream(double imageSizeWidth, double imageSizeHeight);
  public static Stream GetImageStream(IEnumerable<Line> lines, Size imageSize, Color backgroundColor);
}
```

## Usage Syntax
[usage]: #usage-syntax

### XAML Usage

```xml
<views:DrawingView x:Name="DrawingViewControl" 
   BackgroundColor="LightGray" 
   HeightRequest="200"
   DefaultLineColor="Red"
   DefaultLineWidth="3"
   Lines="{Binding Lines, Mode=TwoWay}"
   ClearOnFinish="{Binding Source={x:Reference ClearOnFinish}, Path=IsToggled}"
   MultiLineMode="{Binding Source={x:Reference MultiLineMode}, Path=IsToggled}"/>
```

### C# Usage

```cs
new DrawingView
{
   BackgroundColor = Colors.LightGray,
   HeightRequest = 200,
   DefaultLineColor = Colors.Red,
   DefaultLineWidth = 3
}.Bind(DrawingView.Lines, nameof(ViewModel.Lines))
 .Bind(DrawingView.ClearOnFinish, nameof(ViewModel.ShouldClear))
 .Bind(DrawingView.MultiLineMode, nameof(ViewModel.ShouldAllowMultiLine));
```

## Alternatives
[alternatives]: #alternatives

Might be a duplicate of [`Microsoft.Maui.Graphics`](https://github.com/dotnet/Microsoft.Maui.Graphics)?

## Unresolved Questions
[unresolved]: #unresolved-questions

Is `DrawingView` still necessary? Or has it been superseded by [`Microsoft.Maui.Graphics`](https://github.com/dotnet/Microsoft.Maui.Graphics)?

### PR DESCRIPTION
**Pull request by VladislavAntonyuk**
### Description of Change ###

- Update the background immediately 
- Register Control in Handler
- Set windows min version to 18362 to remove warning
- Ensure user canvas settings don't affect DrawingLine
- Use Background instead of BackgroundColor
- RC2

 ### Linked Issues ###
 

 - Fixes #108 

 ### PR Checklist ###
 
 - [ ] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [ ] Has samples (if omitted, state reason in description)
 - [ ] Rebased on top of `main` at time of PR
 - [ ] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/XYZ 

 ### Additional information ###

### CODE REVIEW

File: src/CommunityToolkit.Maui.Core/Views/DrawingView/PlatformView/MauiDrawingView.shared.cs [hunk 1]
```diff
@@ -57,7 +58,15 @@ public partial class MauiDrawingView
...
 	/// Drawable background
 	/// </summary>
-	public Paint Paint { get; set; } = new SolidPaint(DrawingViewDefaults.BackgroundColor);
+	public Paint Paint
+	{
+		get => paint;
+		set
+		{
+			paint = value;
+			Redraw();
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Views/DrawingView/PlatformView/MauiDrawingView.shared.cs:67]: Fix Background is not redrawn after set

File: src/CommunityToolkit.Maui.Core/Views/DrawingView/PlatformView/MauiDrawingView.shared.cs [hunk 2]
```diff
@@ -202,6 +214,10 @@ static void SetStroke(in ICanvas canvas, in float lineWidth, in Color lineColor)
 		{
 			canvas.StrokeColor = lineColor;
 			canvas.StrokeSize = lineWidth;
+			canvas.StrokeDashOffset = 0;
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Views/DrawingView/PlatformView/MauiDrawingView.shared.cs:217]: explicitly set to default values to make sure user doesn't override it in DrawAction

File: src/CommunityToolkit.Maui.Core/Views/DrawingView/Service/DrawingViewService.windows.cs [hunk 3]
```diff
@@ -82,22 +82,17 @@ static async ValueTask<Stream> GetCanvasRenderTargetStream(CanvasRenderTarget? c
...
-		CanvasDrawingSession session, 
-		ICollection<PointF> points, 
+	static void DrawStrokes(CanvasDrawingSession session, 
+		IEnumerable<PointF> points, 
 		Color lineColor, 
-		float lineWidth)
+		float lineWidth,
+		Size offset)
 	{
-		var minPointX = points.Min(p => p.X);
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Views/DrawingView/Service/DrawingViewService.windows.cs:96]: it was calculated for each line which is not correct.

**Comment by VladislavAntonyuk** [comment]
Sample app: https://github.com/VladislavAntonyuk/MauiSamples/tree/paint

File: src/CommunityToolkit.Maui.Core/Views/DrawingView/PlatformView/MauiDrawingView.shared.cs [hunk 4]
```diff
@@ -67,7 +76,10 @@ public void Initialize()
 #if ANDROID || IOS || MACCATALYST
 		Drawable = new DrawingViewDrawable(this);
 #elif WINDOWS
-		((Microsoft.Maui.Graphics.Win2D.W2DGraphicsView)Content).Drawable = new DrawingViewDrawable(this);
+		if (OperatingSystem.IsWindowsVersionAtLeast(10, 0, 18362))
```
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Views/DrawingView/PlatformView/MauiDrawingView.shared.cs:79]: What will happen if the windows version is below that?

Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Views/DrawingView/PlatformView/MauiDrawingView.shared.cs:79]: we do not see the lines:
image

we see image and events, but DrawingView Drawable is empty (but control still take it's place)

Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Views/DrawingView/PlatformView/MauiDrawingView.shared.cs:79]: invisible ink feature

Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Views/DrawingView/PlatformView/MauiDrawingView.shared.cs:79]: can we add a log message informing the user that he needs to increase the target version?

Comment (brminnick) [reply src/CommunityToolkit.Maui.Core/Views/DrawingView/PlatformView/MauiDrawingView.shared.cs:79]: We could use the `SupportedOSPlatformAttribute` and specify the version number. 

Something like this ensures that a warning appears for all versions of windows older than 10.0.18362.0:
```cs
[UnsupportedOSPlatform("windows")]
[SupportedOSPlatform("windows10.0.18362.0")
```

https://docs.microsoft.com/en-us/dotnet/standard/analyzers/platform-compat-analyzer#advanced-scenarios-for-attribute-combinations

Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Views/DrawingView/PlatformView/MauiDrawingView.shared.cs:79]: I would prefer a simple log message. SupportedOSPlatform attribute generates a warning. Also, it is a MauiDrawingView, which is not used directly in Xaml or C#

File: src/CommunityToolkit.Maui.Core/Views/DrawingView/Service/DrawingViewService.windows.cs [hunk 5]
```diff
@@ -173,4 +170,46 @@ static void DrawInk(this CanvasDrawingSession session, IEnumerable<InkStroke> st
...
+					var gradientStops = new CanvasGradientStop[linearGradientBrush.GradientStops.Length];
+					for (var index = 0; index < linearGradientBrush.GradientStops.Length; index++)
+					{
+						var gradientStop = linearGradientBrush.GradientStops[index];
+						gradientStops[index] = new CanvasGradientStop(gradientStop.Offset, gradientStop.Color.ToWindowsColor());
+					}
+
+					return new CanvasLinearGradientBrush(canvasResourceCreator, gradientStops)
+					{
+						StartPoint = new Vector2((float)(linearGradientBrush.StartPoint.X * size.Width), (float)(linearGradientBrush.StartPoint.Y * size.Height)),
```
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Views/DrawingView/Service/DrawingViewService.windows.cs:191]: why in some places do you use the property `Size.Height` and in others do you use the field `Size._height`? Would be nice to have consistence

**Review by pictos — CHANGES_REQUESTED** [review]
Hey @VladislavAntonyuk I found some issues on Android, let me know if you need any help on those.

On Android, I found this issue when pressing the `Generate Image with Random Lines` button
image

Also, the line can be drawn outside the DrawingView bounds
image

Also, I can't see the `Current gesture pad image`, as you can see in the screenshot below

image

Another thing, with the use set to true `ShouldClearOnFinish` and `IsMultiLineModeEnabled`, the DrawingView deletes the line, working as `IsMultiLineModeEnabled = false`, we should document that to avoid confusion. Because imagine that I want to use this control as a signature pad, to me the `Finish` will be when the user doesn't draw anything more after 5s idle.

File: src/CommunityToolkit.Maui.Core/Views/DrawingView/Service/DrawingViewService.windows.cs (hunk 5 above)
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Views/DrawingView/Service/DrawingViewService.windows.cs:191]: To avoid casting. Height is double, _height is float.

**Comment by VladislavAntonyuk** [comment]
As for the ClearOnFinish and MultiLineMode, it is documented. 
ClearOnFinish is true + MultiLineMode is true = MultiLineMode is false
We provide a method Clean, so users can automatically clean the canvas after a small delay.

**Comment by pictos** [comment]
> As for the ClearOnFinish and MultiLineMode, it is documented. ClearOnFinish is true + MultiLineMode is true = MultiLineMode is false We provide a method Clean, so users can automatically clean the canvas after a small delay.

So that is just me being pepega

**Review by pictos — APPROVED** [review]
From my tests everything looks good! 
Thanks @VladislavAntonyuk

**Pull request merged**